pkg/recommendation/map.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2020 The MayaData Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package recommendation
+
+import (
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"mayadata.io/cstorpoolauto/types"
+	bdutil "mayadata.io/cstorpoolauto/util/blockdevice"
+)
+
+/*
+This function builds a map using existing block devices. Block devices are
+grouped by device kind ie - HDD or SSD and in one group all the block devices
+are arranged by node name.
+
+1. initial stage it will be a list of block devices.
+[]blockdevice
+
+2.
+at last it will be converted into below map -
+- HDD:
+	- node-1 []blockdevice
+	- node-2 []blockdevice
+	- node-3 []blockdevice
+- SSD:
+	- node-1 []blockdevice
+	- node-2 []blockdevice
+	- node-3 []blockdevice
+
+3. In some case all the may not have block device or all the
+types of block device. For that this may look like this -
+- HDD:
+	- node-1 []blockdevice
+	- node-3 []blockdevice
+	- node-4 []blockdevice
+	- node-7 []blockdevice
+- SSD:
+	- node-3 []blockdevice
+	- node-5 []blockdevice
+	- node-7 []blockdevice
+*/
+func buildNodeBlockDeviceMap(
+	bdlist unstructured.UnstructuredList) map[string]map[string][]BlockDeviceInfo {
+	deviceTypeNodeBlockDeviceMap := make(map[string]map[string][]BlockDeviceInfo)
+
+	for _, bd := range bdlist.Items {
+
+		// If block device is not in active state or we got any error while
+		// fetching block device status then we can not use that block device
+		// to create a pool.
+		if ok, err := bdutil.IsActiveOrError(bd); err != nil || !ok {
```

Comment:
Sorry for giving this comment now should have been given in the previous PR only...
The name of the method here i.e., **IsActiveOrError** should/could be only **IsActive** which means that it is checking if the bd is active or not which will be conveyed by the boolean value (ok) and if there is some error it would be obviously in the err variable.

**IsActiveOrError** somehow gave me a feeling that the bool value could be true in case of an error also.

Reply:
Do we want some logs in error cases may be at higher verbose levels?

Reply:
@sagarkrsd we are trying to follow IsActive & IsActiveOrError methods if both forms are required.
First one can only return bool, second one can return error as second arg.
However, if we find that all requirements suffice with one method then we can have IsActive that returns bool & error.
cc @shovanmaity

Reply:
Alternatively we can have IsActiveOrError return only error. This implies error is returned for error cases as well as for cases when device is not active. This should purely be done based on the requirements.

Reply:
In these particular cases, since we are swallowing the error, we can only use IsActive().
Now we need to be sure if we should be swallowing the error?

Reply:
Please check if we should avoid swallowing errors

Reply:
Thank you @sagarkrsd  for raising your concerns and put your thoughts. :+1: 

We can have a discussion (with you @AmitKumarDas  and @avishnu and anyone who is interested) on where to use **Error** and where not some standard or good practices. I will take action(update function name) after that discussion.

I will not make any changes related to this(if required after discussion) in this PR.
That will be in some other PR.  Thanks again for raising your concerns and put your thoughts.

Reply:
find the discussion here - https://mayadata-team.slack.com/archives/CT2DK8329/p1582039023276800